numpyro/svi.py

```diff
@@ -177,17 +171,9 @@ def elbo(param_map, model, guide, model_args, guide_args, kwargs, constrain_fn,
     """
     param_map = constrain_fn(param_map)
     guide_log_density, guide_trace = log_density(guide, guide_args, kwargs, param_map)
-    if is_autoguide:
-        # in autoguide, a site's value holds intermediate value
-        for name, site in guide_trace.items():
-            if site['type'] == 'sample':
-                param_map[name] = site['value']
-    else:
-        # NB: we only want to substitute params not available in guide_trace
-        param_map = {k: v for k, v in param_map.items() if k not in guide_trace}
-        model = replay(model, guide_trace)
-    model_log_density, _ = log_density(model, model_args, kwargs, param_map,
-                                       skip_dist_transforms=is_autoguide)
+    param_map = {k: v for k, v in param_map.items() if k not in guide_trace}
+    model = replay(model, guide_trace)
+    model_log_density, _ = log_density(model, model_args, kwargs, param_map)
```

Comment:
My understanding was that if the same sample site is present in both, they both will have a constant log det term for scaling which should cancel out when we do `model_log_density - guide_log_density`. We can probably also include it for both. Let me know if I got that wrong.

Reply:
I think that we do inference for SVI in constrained space, so we have to compute `log_prob` of constrained values. The issue will happen when prior is `dist.Normal` but guide is `dist.Uniform`, or when prior and guide are transformed distributions with different transforms. In SVI, we compute KL of 2 constrained variables, while in autoguide, we compute KL of 2 unconstrained variables.

Reply:
> The issue will happen when prior is dist.Normal but guide is dist.Uniform

This will not work if the guide has a different support for a sample site than the model. I have been assuming that any well designed guide should retain the sample site's support (like we do in autoguides). So even if we use a uniform distribution in the guide, we would probably want to transform it back to the Real domain so as to map to the support of the site in the model (a Normal distribution). We can add a check for this, if needed.

> or when prior and guide are transformed distributions with different transforms. 

Could you give an example? Suppose we have a uniform distribution from [0, 5] in the model and we have a transformed distribution in the guide that transforms from a Normal distribution to [0, 5]. Then I think that in both the cases we can just compute the log density for the base values (ignoring the scaling term from [0, 1] -> [0, 5]).

Maybe we can add some tests to compute the elbo for all of these cases, but just want to make sure I really understand the issues involved, and am not oversimplifying.

Reply:
My assumption is for the case `model` and `guide` have different transforms, such as
```
def model():
    sample('x', dist.TransformedDistribution(dist.Normal(), AffineTransform(0, 1))

def guide():
    sample('x', dist.TransformedDistribution(dist.Normal(), AffineTransform(0, 2))
```

Because `AffineTransform(0, 1)` logdet is different from `AffineTransform(0, 2)` logdet, `log_prob_guide(x) - log_prob_model(x)` will be different from `log_prob_guide(base_x) - log_prob_model(base_x)`.

Similarly, for bounded support, logdet of `Sigmoid` is different from logdet of `Compose([Tanh, Affine(0.5, 0.5))` though both of them have the same codomain `[0, 1]`.

Reply:
Thanks for explaining, that makes sense. I am still unclear about why we need to do anything differently for autoguides though. What goes wrong if we instead have `skip_dist_transforms=False` for both guide and model under all cases?

Reply:
That passes all our tests btw - the computation of the elbo loss term is different in `test_autoguides.py::test_dynamic_supports` but that's expected since the log density term will be different for `Uniform(0, alpha)` and `Uniform(0, 1)` but it shouldn't affect the optimal parameter values.

Reply:
I have reverted the autoguide check. While it would be nice to not have to make a distinction in svi, I don't see any easy way to accomplish this currently.